File: CTrecon_example/ct_recon.m
clear all; close all;

xd = 128;

figure(1);
true_object = abs(phantom(xd));

subplot(2,3,1), imshow(true_object, [0 max(true_object(:))]);

azi_angles = [1:1:180];

meas_data = radon(true_object,azi_angles);

subplot(2,3,2), imshow(meas_data', [0 max(meas_data(:))]);

rec = ones(size(true_object)); % initial reconstruction image



sinogram_ones = ones(size(meas_data));
sens = iradon(sinogram_ones,azi_angles,'none',xd);

    subplot(2,3,4), imshow(rec, [0 max(rec(:))]);
   
pause
for it = 1:10
    fp = radon(rec,azi_angles);
        subplot(2,3,5), imshow(fp', [0 max(fp(:))]);
    ratio = meas_data ./ (fp + 0.0001);
    subplot(2,3,3), imshow(ratio', [0 max(ratio(:))]);
    bp_ratio = iradon(ratio,azi_angles,'none',xd);
    
        subplot(2,3,6), imshow(bp_ratio, [0 max(rec(:))]);
    rec = rec .* bp_ratio

    pause
    subplot(2,3,4), imshow(rec, [0 max(rec(:))]);


end    

File: CTrecon_example_python/CTrecon.py
#Simple iterative image reconstruction algorithm
#ML-EM(Maximum Likelihodd - expectaion maximisation)

from turtle import circle
from skimage.data import shepp_logan_phantom
from skimage.transform import radon,iradon, rescale
import numpy as np
import matplotlib.pyplot as plt


plt.ion()
activity_level = 0.1;

true_object = shepp_logan_phantom();
true_object = rescale(activity_level * true_object, 0.5)


fig, axs = plt.subplots(2,3,figsize = (20,10))
axs[0,0].imshow(true_object, cmap = 'Greys_r');     axs[0,0].set_title('object')


#generate simulated sinogram data
azi_angles = np.linspace(0.0,180.0,180, endpoint = False)
sinogram = radon(true_object, azi_angles , circle = False)
axs[0,1].imshow(sinogram.T, cmap = 'Greys_r');      axs[0,1].set_title('Sinogram')





mlem_rec = np.ones(true_object.shape); # iteraiton 0 (k = 0)
sino_ones = np.ones(sinogram.shape)
sens_image = iradon(sino_ones,azi_angles,circle = False,filter_name = None ) 
for iter in range(20):
    fp = radon(mlem_rec,azi_angles,circle = False) # forward projection of mlem_rec at iteration k A x^k
    ratio = sinogram / (fp +0.000001) # ratio sinogram
    correction = iradon(ratio,azi_angles,circle = False, filter_name = None ) / sens_image

    axs[1,0].imshow(mlem_rec, cmap = 'Greys_r');    axs[0,1].set_title('mlem recon')
    axs[1,1].imshow(fp.T, cmap = 'Greys_r');     axs[1,1].set_title('FP of recon')
    axs[0,2].imshow(ratio.T, cmap = 'Greys_r');     axs[0,2].set_title('Ratio Sinogram')



    axs[1,2].imshow(correction, cmap = 'Greys_r'); axs[1,2].set_title('BP of ratio ')
    mlem_rec = mlem_rec * correction
    axs[1,0].imshow(mlem_rec, cmap = 'Greys_r'); axs[1,0].set_title('MLEM Reconstructed image IT = %d' % (iter+1))
    plt.show()
    plt.pause(0.5)

plt.show(block = True)